crates/trie/trie/src/hashed_cursor/post_state.rs

```diff
@@ -25,141 +25,184 @@ where
+        = HashedPostStateCursor<'overlay, CF::StorageCursor<'cursor>, U256>
     where
         Self: 'cursor;
 
     fn hashed_account_cursor(&self) -> Result<Self::AccountCursor<'_>, DatabaseError> {
         let cursor = self.cursor_factory.hashed_account_cursor()?;
-        Ok(HashedPostStateAccountCursor::new(cursor, &self.post_state.as_ref().accounts))
+        Ok(HashedPostStateCursor::new(Some(cursor), &self.post_state.as_ref().accounts))
     }
 
+    /// Returns a `HashedPostStateCursor` for storage slots that merges:
+    /// 1. `post_state_cursor` - In-memory overlay of storage updates (None if address has no
+    ///    updates)
+    /// 2. `cursor` - DB cursor for existing storage (None if storage was wiped via `SELF-DESTRUCT`)
+    ///
+    /// When storage is wiped, the DB cursor is omitted since all previous storage is destroyed.
     fn hashed_storage_cursor(
         &self,
         hashed_address: B256,
     ) -> Result<Self::StorageCursor<'_>, DatabaseError> {
-        let cursor = self.cursor_factory.hashed_storage_cursor(hashed_address)?;
-        Ok(HashedPostStateStorageCursor::new(
-            cursor,
-            self.post_state.as_ref().storages.get(&hashed_address),
-        ))
+        static EMPTY_UPDATES: Vec<(B256, Option<U256>)> = Vec::new();
+
+        let post_state_storage = self.post_state.as_ref().storages.get(&hashed_address);
+        let (storage_slots, wiped) = post_state_storage
+            .map(|u| (u.storage_slots_ref(), u.is_wiped()))
+            .unwrap_or((&EMPTY_UPDATES, false));
+
+        let cursor = if wiped {
+            None
+        } else {
+            Some(self.cursor_factory.hashed_storage_cursor(hashed_address)?)
+        };
+
+        Ok(HashedPostStateCursor::new(cursor, storage_slots))
     }
 }
 
-/// The cursor to iterate over post state hashed accounts and corresponding database entries.
-/// It will always give precedence to the data from the hashed post state.
+/// A cursor to iterate over state updates and corresponding database entries.
+/// It will always give precedence to the data from the post state updates.
 #[derive(Debug)]
-pub struct HashedPostStateAccountCursor<'a, C> {
-    /// The database cursor.
-    cursor: C,
-    /// Forward-only in-memory cursor over accounts.
-    post_state_cursor: ForwardInMemoryCursor<'a, B256, Account>,
-    /// Reference to the collection of account keys that were destroyed.
-    destroyed_accounts: &'a B256Set,
-    /// The last hashed account that was returned by the cursor.
+pub struct HashedPostStateCursor<'a, C, V> {
+    /// The underlying `database_cursor`. If None then it is assumed there is no DB data.
+    cursor: Option<C>,
+    /// Entry that `database_cursor` is currently pointing to.
+    cursor_entry: Option<(B256, V)>,
+    /// Forward-only in-memory cursor over underlying V.
+    post_state_cursor: ForwardInMemoryCursor<'a, B256, Option<V>>,
```

Comment:
I would introduce a trait like:

```
trait HashedPostStateCursorValue<V> {
    fn as_option(&self) -> Option<&'_ V>
}
```

And implement that for both `Option<Account>` and `U256` here. Then that trait can be used in the ForwardInMemoryCursor instead of `Option<V>`. This way we can don't need to use `Option<U256>` on the HashedStorageSorted.